src/meta-srv/src/region/supervisor.rs

```diff
@@ -0,0 +1,523 @@
+                #[cfg(test)]
+                Event::Dump(sender) => {
+                    let _ = sender.send(self.failure_detector.dump());
+                }
+            }
+        }
+        info!("RegionSupervisor is stopped!");
+    }
+
+    async fn handle_region_failures(&self, mut regions: Vec<(ClusterId, DatanodeId, RegionId)>) {
+        if regions.is_empty() {
+            return;
+        }
+        match self.is_maintenance_mode().await {
+            Ok(false) => {}
+            Ok(true) => {
+                info!("Maintenance mode is enabled, skip failover");
+                return;
+            }
+            Err(err) => {
+                error!(err; "Failed to check maintenance mode");
+                return;
+            }
+        }
+
+        let migrating_regions = regions
+            .extract_if(|(_, _, region_id)| {
+                self.region_migration_manager.tracker().contains(*region_id)
+            })
+            .collect::<Vec<_>>();
+
+        for (cluster_id, datanode_id, region_id) in migrating_regions {
+            self.failure_detector
+                .remove(&(cluster_id, datanode_id, region_id));
+        }
+
+        warn!(
+            "Detects region failures: {:?}",
+            regions
+                .iter()
+                .map(|(_, datanode, region)| (datanode, region))
+                .collect::<Vec<_>>()
+        );
+        for (cluster_id, datanode_id, region_id) in regions {
+            match self.do_failover(cluster_id, datanode_id, region_id).await {
+                Ok(_) => self
+                    .failure_detector
+                    .remove(&(cluster_id, datanode_id, region_id)),
+                Err(err) => {
+                    error!(err; "Failed to execute region failover for region: {region_id}, datanode: {datanode_id}");
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn is_maintenance_mode(&self) -> Result<bool> {
+        self.kv_backend
+            .exists(MAINTENANCE_KEY.as_bytes())
+            .await
+            .context(error::KvBackendSnafu)
+    }
+
+    async fn do_failover(
+        &self,
+        cluster_id: ClusterId,
+        datanode_id: DatanodeId,
+        region_id: RegionId,
+    ) -> Result<()> {
+        let from_peer = self
+            .peer_lookup
+            .datanode(cluster_id, datanode_id)
+            .await
+            .context(error::LookupPeerSnafu {
+                peer_id: datanode_id,
+            })?
+            .context(error::PeerUnavailableSnafu {
+                peer_id: datanode_id,
+            })?;
+        let mut peers = self
+            .selector
+            .select(
+                cluster_id,
+                &self.selector_context,
+                SelectorOptions {
+                    min_required_items: 1,
+                    allow_duplication: false,
+                },
+            )
+            .await?;
+        let to_peer = peers.remove(0);
```

Comment:
Should we check the length of peers to avoid potential panic?

Also we can do `peers.pop()` to avoid shifting remaining elements in peers if we don't care about the order of `to_peer`.

Reply:
The `Selector` ensures it always returns `min_required_items` peers.